File: experiments/03_synth_data_generation/c_src/batch06_bits.c
/* Batch 06: Bit manipulation — pure functions on unsigned integers */

int is_power_of_two(unsigned int n) {
    return n != 0 && (n & (n - 1)) == 0;
}

unsigned int count_set_bits(unsigned int n) {
    unsigned int count = 0;
    while (n) {
        count += n & 1;
        n >>= 1;
    }
    return count;
}

unsigned int parity(unsigned int n) {
    unsigned int p = 0;
    while (n) {
        p ^= (n & 1);
        n >>= 1;
    }
    return p;
}

unsigned int reverse_bits(unsigned int n) {
    unsigned int result = 0;
    for (int i = 0; i < 32; i++) {
        result = (result << 1) | (n & 1);
        n >>= 1;
    }
    return result;
}

unsigned int next_power_of_two(unsigned int n) {
    if (n == 0) return 1;
    n--;
    n |= n >> 1;
    n |= n >> 2;
    n |= n >> 4;
    n |= n >> 8;
    n |= n >> 16;
    return n + 1;
}

unsigned int lowest_set_bit(unsigned int n) {
    return n & (~n + 1);
}

unsigned int clear_lowest_set_bit(unsigned int n) {
    return n & (n - 1);
}

File: experiments/03_synth_data_generation/py_src/batch06_bits.py
"""Batch 06: Bit manipulation — pure functions on non-negative integers."""

def is_power_of_two(n: int) -> bool:
    return n != 0 and (n & (n - 1)) == 0

def count_set_bits(n: int) -> int:
    count = 0
    while n:
        count += n & 1
        n >>= 1
    return count

def parity(n: int) -> int:
    p = 0
    while n:
        p ^= (n & 1)
        n >>= 1
    return p

def reverse_bits_32(n: int) -> int:
    result = 0
    for _ in range(32):
        result = (result << 1) | (n & 1)
        n >>= 1
    return result

def next_power_of_two(n: int) -> int:
    if n == 0:
        return 1
    n -= 1
    n |= n >> 1
    n |= n >> 2
    n |= n >> 4
    n |= n >> 8
    n |= n >> 16
    return n + 1

def lowest_set_bit(n: int) -> int:
    return n & (~n + 1)

def clear_lowest_set_bit(n: int) -> int:
    return n & (n - 1)
